### PR TITLE
fix(definitions): optimize and fix definition highlighting

### PR DESCRIPTION
**Pull request by steelsojka**
This will fix https://github.com/nvim-treesitter/nvim-treesitter/issues/282

This will also improve definition performance on successive definition and usage lookups.

### CODE REVIEW

File: lua/nvim-treesitter/locals.lua [hunk 1]
```diff
@@ -158,6 +140,42 @@ function M.recurse_local_nodes(local_def, accumulator, full_match, last_match)
   end
 end
 
+--- Get a single dimension table to look definition nodes by scope and text.
+-- This is memoized by buffer tick to avoid computing on every lookup.
+M.get_defintions_lookup_table = ts_utils.memoize_by_buf_tick(function(bufnr)
```
Comment (vigoux) [inline lua/nvim-treesitter/locals.lua:145]: Little typo here

Comment (vigoux) [reply lua/nvim-treesitter/locals.lua:145]: Furthermore, I don't see the point of exposing that if we don't expose the `get_definition_id` too.

Comment (vigoux) [reply lua/nvim-treesitter/locals.lua:145]: Last comment I promise, can you add a bit more doc regarding what represents the return value of this function, and maybe the idea behind it ? I don't find that especially easy to get.

Comment (steelsojka) [reply lua/nvim-treesitter/locals.lua:145]: Yeah good point. I will update this.

Comment (theHamsta) [reply lua/nvim-treesitter/locals.lua:145]: and it has a typo

File: lua/nvim-treesitter/locals.lua [hunk 2]
```diff
@@ -24,6 +24,16 @@ function M.get_locals(bufnr)
...
 end
 
+--- Creates unique id for a node based on text and range
+-- @param scope: the scope node of the definition
+-- @param bufnr: the buffer
+-- @param node_text: the node text to use
+-- @returns a string id
+function M.get_definition_id(scope, node_text)
+  -- Add a vaild starting character in case node text doesn't start with a valid one.
+  return table.concat({ 'k', node_text, scope:range() }, '_')
```
Comment (theHamsta) [inline lua/nvim-treesitter/locals.lua:34]: why not node.id? Is node.id not unique when compared over time?

**Comment by theHamsta** [comment]
Nice work. This is no fast enough to run on `CursorMoved`. Maybe we can offer that as an option?

File: lua/nvim-treesitter/locals.lua (hunk 2 above)
Comment (steelsojka) [reply lua/nvim-treesitter/locals.lua:34]: Honestly I didn't even nodes had IDs. I don't think I saw it in the docs. I'm also not 100% sure if the node id will be the same across different trees...

Comment (theHamsta) [reply lua/nvim-treesitter/locals.lua:34]: I guess the ID should change on reparsing. I am however not sure. You could  test it.

**Pull request merged**